pytorch_lightning/utilities/trainer_cli.py

```diff
@@ -0,0 +1,223 @@
+
+        Args:
+            datamodule_class: A LightningDataModule class.
+            nested_key: Name of the nested namespace to store arguments.
+        """
+        assert issubclass(datamodule_class, LightningDataModule)
+        self.add_class_arguments(datamodule_class, nested_key)
+
+
+class SaveConfigCallback(Callback):
+    """Saves a LightningCLI config to the log_dir when training starts"""
+
+    def __init__(self, parser, config):
+        self.config_dump = parser.dump(config, skip_none=False)
+
+    def on_train_start(self, trainer, pl_module):
+        config_path = os.path.join(trainer.logger.log_dir, 'config.yaml')
+        with open(config_path, 'w') as outstream:
+            outstream.write(self.config_dump)
+
+
+class LightningCLI:
+    def __init__(
+        self,
+        model_class: Type[LightningModule],
+        datamodule_class: Type[LightningDataModule] = None,
+        save_config_callback: Type[Callback] = SaveConfigCallback,
+        trainer_class: Type[Trainer] = Trainer,
+        description: str = 'pytorch-lightning trainer command line tool',
+        parse_env: bool = False,
+        **kwargs
+    ):
+        """
+        Implementation of a simple configurable Trainer command line tool
+
+        Receives as input pytorch-lightning classes, which are instantiated using a
+        parsed configuration file or command line args and then runs trainer.fit.
+
+        Example, first implement the trainer.py tool as::
+
+            from mymodels import MyModel
+            from pytorch_lightning.utilities.jsonargparse_utils import LightningCLI
+            LightningCLI(MyModel)
+
+        Then in a shell, run the tool with the desired configuration::
+
+            $ python trainer.py --print-config > config.yaml
+            $ nano config.yaml  # modify the config as desired
+            $ python trainer.py --cfg config.yaml
+
+        Args:
+            model_class: The LightningModule class to train on.
+            datamodule_class: An optional LightningDataModule class.
+            save_config_callback: A callback class to save the training config.
+            trainer_class: An optional extension of the Trainer class.
+            description: Description of the tool shown when running --help.
+            parse_env: Whether environment variables are also parsed.
+            **kwargs: Additional arguments to instantiate Trainer.
+        """
+        if 'callbacks' not in kwargs:
+            kwargs['callbacks'] = []
+
+        self.model_class = model_class
+        self.datamodule_class = datamodule_class
+        self.save_config_callback = save_config_callback
+        self.trainer_class = trainer_class
+        self.trainer_kwargs = kwargs
+
+        self.init_parser(description, parse_env)
+        self.add_arguments_to_parser(self.parser)
+        self.add_core_arguments_to_parser()
+        self.parse_arguments()
+        self.instantiate_classes()
```

Comment:
Adding a hook after parsing arguments and before instantiation can help with situations where there are callbacks which are only added to the trainer as needed, based on command line arguments. For example, LearningRateMonitor and EarlyStopping. These can also be addressed by overriding instantiate_classes, but I feel that a hook for processing that parsed arguments is cleaner for that. 

It could even be a dedicated hook, e.g. add_callbacks which gets the callbacks list and config as arguments and returns an updated list. Or it could be similar to before/after init, which are more open ended.

Reply:
Maybe the list of callbacks should be given to LightningCLI. This way it is possible to also make the callbacks configurable with the same config file, instead of hardcoding things in the code. So for example in the config file we could have:
```
callbacks:
  LearningRateMonitor:
    log_momentum: true
  EarlyStopping:
    patience: 5
trainer:
  ...
```

Reply:
That's an interesting idea. But doing it like that would remove the dependency between the args and the callbacks. For example, I have no wish to monitor learning rate if the model args specify that the learning rate should be constant.

Also, you won't be able to pass custom callbacks written by the user as part of the config file, right? So some hard-coding can be expected.

I think that being able to specify callbacks or callback params based on args can be important. So, if a way to incorporate it into the process, such as a hook, is possible I'd recommend adding it.

Reply:
See comment https://github.com/PyTorchLightning/pytorch-lightning/pull/4492#issuecomment-735705480

Reply:
Certainly this does not exclude having a hook for other use cases like the one you mentioned.

Reply:
I have added `before_instantiate_classes` method to the cli class.